### PR TITLE
Added Support for Msg91 phone provider

### PR DESCRIPTION
**Pull request by MeetChaudhari**
## What kind of change does this PR introduce?

This PR adds support for Msg91.com , One of the best highly reliable Indian SMS provider rates even lesser than textlocal.

## What is the current behavior?

Actually currently present textlocal integration for Indian audience highly unreliable not from the integration point, its from the textlocal servers and  my company is currently developing an huge artist centric platform using supabase during our testing we had lot of issues due to textlocal. So today i have added Msg91 support

## Additional context

The implementation is loosely based on textlocal implementation. 
And due to Indian Government mandate, the message template, header has to be registered before sending any sms. 
 
Also my implementation takes 3 params 

- Msg91 authkey
- DLT template Id
- Sender Id

For above this i will add documentation for obtain each of the params

### CODE REVIEW

**Comment by kangmingtay** [comment]
Hey @MeetChaudhari MeetChaudhari, thanks for the contribution! Unfortunately, i don't think we plan to merge this in since the team has decided to go with webhooks in the future for custom sms providers which would be more scalable in terms of maintaining the codebase. However, if you need this feature urgently, we recommend that you fork the codebase to include your changes in it.

**Comment by MeetChaudhari** [comment]
@kangmingtay, I understand and agree with your decision, but the thing is, we do not want to self-host the platform, as such we have a lot of microservices to deal with and as founder, I need a bit of stability in the signup process that's the reason I added this like if it is possible to please help me out I would be very grateful to your team and whenever you guys switch to webhooks I'll do all the necessary changes required for my contribution. 

Like till you guys implement the webhook solution, it would very helpful for me to get started for me with my pr.

**Comment by kiwicopple** [comment]
Hey @MeetChaudhari - thanks for the additional context. 

it's a tricky balance we need to strike between accepting a lot of PRs now (delaying the webhooks implementation) and commiting to the webhook approach.

I understand that `textlocal` might not be as cheap as `Msg91`, but it's already implemented and available. Is there any reason that you can't use it, beyond cost?

**Comment by MeetChaudhari** [comment]
hey @kiwicopple , actually it's not about the cost it is about the reliability as a textlocal customer I have already purchased a good number of sms credits on their platform the problem is sometimes it works flawlessly sometimes it just gives random errors on very frequent basis and this is not an issue with the supabase textlocal integration authored by devkiran, it's from textlocal end. I even had a discussion with devkiran about the issue with textlocal.

And just think if I want to get my dream project into production I seriously need to have the reliability of a service.

**Comment by kangmingtay** [comment]
Hey @MeetChaudhari, if cost is not an issue and you're looking for reliability, will it be possible to use one of the other providers instead? We also support using Twilio, Messagebird and Vonage on Supabase. Is there a particular reason why you are choosing Msg91 over these other providers?

**Comment by MeetChaudhari** [comment]
Cause that becomes comparatively too expensive for us, as our platform is situated in India for comparison

This are rates for sending one message in India in USD. (All prices are taken from their respective brand sites.)

- Vonage               $0.0436
- MessageBird      $0.0420
- Twilio                  $0.0490

- TextLocal.          $0.0036    29.5 paisa (INR)
- Msg91                $0.0030    25.0 paisa (INR)

so its pretty obvious for us to use Indian providers

**Comment by kiwicopple** [comment]
> *problem is sometimes it works flawlessly sometimes it just gives random errors on very frequent basis*

I guess the challenge for us here - how do we know that it's not just the same problem with msg91? From a quick search, there are quite a few SMS providers in India. If we keep merging new providers and each of them is buggy, it creates more delays for the Auth team to implement the "right" solution (webhooks). 

Have you raised a ticket with the textlocal team to help solve any issues?

**Comment by MeetChaudhari** [comment]
Hey @kiwicopple i agree with you, the problem was i tried everything possible even my textlocal logs, i agree did not raise any tickets with textlocal team, but not in case of this integration, i signed up as startup on msg91, i have prior experience with msg91, this company serves for multiple unicorn companies in india and my prior experience is flawless with them, supabase textlocal Integration made me switch to textlocal.

And as i signed up as startup which is what i have, i have a contact/ direct support of developer form msg91 to mitigate any issues, like even while making integration i had some doubts which was instantly solved by msg91 team.

**Comment by MeetChaudhari** [comment]
https://www.fonada.com/blog/top-10-bulk-sms-service-providers-in-india/

Check this recent article. For more reference

**Comment by kangmingtay** [comment]
@MeetChaudhari We see where you're coming from but we are still hesitant to push forward on this for now because providing top-notch support for this feature could be challenging given the issues we've experienced with sending SMSes from other service providers (TextLocal). For example, it was impossible for our support team to debug issues with TextLocal since there's a requirement for one to own a SIM card from India, which requires one to be a citizen of India. Additionally, if we merge in this PR, it could make it harder to turn down future SMS providers since there may always be a case for another one being more reliable or cost-effective. We also haven't received any other request to add Msg91 as an SMS provider and we don't add integrations based on one-off requests. For example, we keep a list of external OAuth provider requests [here](https://github.com/supabase/gotrue/issues/451) and we only accept contributions for those that are deemed as "in-demand" from our community. 

Our goal is to make gotrue more flexible by introducing features like webhooks, which will allow developers to integrate any third-party software they choose, instead of building a customised solution for every auth flow. 

One alternative that you can consider here will be to host a fork of GoTrue on providers like Fly, which have a free tier, until we've worked out the webhook implementation.

**Comment by MeetChaudhari** [comment]
Yes i completely understand, but as i have talked with msg91 team personally for your problems, they have responded me this

"We can even help their team in testing and resolving any bugs and I can personally be in touch with them for that.
They will not have to get an indian number to debug and do the testing, we'll assist them in that dedicatedly."

And i will take care of documentation and maintenance of the integration till you guys implement the webhook solution with being in direct contact with msg91 team. They will help your team to test out integration perfectly. As mentioned above.

MSG91 team provides three channels for support to all their customers - 
1. Email - support@msg91.com
2. Phone - +918818888733
3. Chat - Available on Website

Specifically for Supabase team, they have agreed to assign a dedicated partner manager which can 
 help you in debugging all issues pertaining to this integration. Details of the Partner Manager is mentioned below along with their direct email and number.

Hardik Menghani
tiger@msg91.com / partners@msg91.com
+918889500122

Hardik can help you in all challenges - may it be with regards to tech/business/regulation.
They can also assist you in debugging issues without getting an Indian Mobile number and for any other queries which may arise in the future.

No other SMS provider in India can offer you services at such a cost-effective pricing with multi-channel support like what MSG91 is offering.

File: .github/workflows/test-m.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
+name: Test
```
Comment (hf) [inline .github/workflows/test-m.yml:1]: Hey why is this file necessary?

File: internal/api/sms_provider/msg91.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	"If your current route is disabled, kindly select another route":  602,
+	"This sender ID is blacklisted, please use a different sender ID": 603,
+	"Please enter at least one correct number to send an SMS":         604,
+	"The scheduled date cannot be more than three weeks":              606,
+	"Please enter the campaign name":                                  607,
+	"Scheduled SMS cannot be less than the current end time":          608,
+}
+
+const (
+	defaultMsg91ApiBase = "https://api.msg91.com/api/sendhttp.php?pluginsource=68"
```
Comment (hf) [inline internal/api/sms_provider/msg91.go:50]: Is this supposed to have `?pluginsource=68`? Any links to their docs would be highly appreciated.

File: internal/api/sms_provider/msg91.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+type Msg91Provider struct {
+	Config  *conf.Msg91ProviderConfiguration
+	APIPath string
+}
+
+type Msg91Response struct {
+	Data string `json:"data"`
+}
+
+// Creates a SmsProvider with the Msg91 Config
```
Comment (hf) [inline internal/api/sms_provider/msg91.go:62]:
```suggestion
// NewMsg91Provider creates a new SmsProvider for Msg91.
```

File: internal/api/sms_provider/msg91.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+		"sender":    {t.Config.SenderId},
+		"DLT_TE_ID": {t.Config.DltTemplateId},
+		"mobiles":   {phone},
+		"message":   {message},
+		"route":     {strconv.Itoa(4)},
+	}
+
+	client := &http.Client{Timeout: defaultTimeout}
+
+	r, err := http.NewRequest("POST", t.APIPath, strings.NewReader(body.Encode()))
```
Comment (hf) [inline internal/api/sms_provider/msg91.go:96]: Please encode the body in a `bytes.NewBuffer` as it is somewhat safer than `NewReader`.

File: internal/api/sms_provider/msg91.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
```
Comment (hf) [inline internal/api/sms_provider/msg91.go:97]:
```suggestion
```

File: internal/api/sms_provider/msg91.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+	if err != nil {
+		return err
+	}
+
+	r.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	res, err := client.Do(r)
+
```
Comment (hf) [inline internal/api/sms_provider/msg91.go:105]:
```suggestion
```

File: internal/api/sms_provider/msg91.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+	if err != nil {
+		return err
+	}
+
+	resBody, _ := io.ReadAll(res.Body)
+
+	resBodyString := string(resBody)
+	resBodyString = strings.Replace(resBodyString, ".", "", -1)
```
Comment (hf) [inline internal/api/sms_provider/msg91.go:113]: Please make sure you handle HTTP status codes before reading the body. I would appreciate it if you could add a short comment explaining the format of the response as I'm not clear why `.` should be replaced with an empty string?

File: internal/api/sms_provider/msg91.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+	if value, ok := err_codes[resBodyString]; ok {
```
Comment (hf) [inline internal/api/sms_provider/msg91.go:115]: Please use `camelCase` for variables and constants.

File: internal/api/sms_provider/msg91.go [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+		return fmt.Errorf("Msg91 error: %v (code: %v)", resBodyString, value)
```
Comment (hf) [inline internal/api/sms_provider/msg91.go:116]: Please also encode the HTTP status code and other information if necessary.

File: internal/conf/configuration.go [hunk 10]
```diff
@@ -490,6 +497,19 @@ func (t *TextlocalProviderConfiguration) Validate() error {
...
+		return errors.New("missing Msg91 Auth key")
+	}
+	if t.DltTemplateId == "" {
+		return errors.New("missing Msg91 Template Id")
+	}
+	if t.SenderId == "" {
+		return errors.New("missing Msg91 Sender Id")
+	}
+	return nil
+}
```
Comment (hf) [inline internal/conf/configuration.go:511]: It would be helpful, if possible, to check the format of these values.

File: .github/workflows/test-m.yml (hunk 1 above)
Comment (MeetChaudhari) [reply .github/workflows/test-m.yml:1]: this was me just checking, the test specified in your repo, cause i wanted to make sure my implementation passes tests

File: internal/api/sms_provider/msg91.go (hunk 2 above)
Comment (MeetChaudhari) [reply internal/api/sms_provider/msg91.go:50]: as i have developed it with msg91 team hand in hand , they asked me to add this flag just for analytics usage, like they can know how many of their users are from supabase. If you say i can ask them to drop you official mail from the msg91 team confirming the same.

File: internal/api/sms_provider/msg91.go (hunk 3 above)
Comment (MeetChaudhari) [reply internal/api/sms_provider/msg91.go:62]: ok i will do the changes

File: internal/api/sms_provider/msg91.go (hunk 4 above)
Comment (MeetChaudhari) [reply internal/api/sms_provider/msg91.go:96]: okay i will do necessary changes, soon.

File: internal/api/sms_provider/msg91.go (hunk 7 above)
Comment (MeetChaudhari) [reply internal/api/sms_provider/msg91.go:112]: ok ill add comments.

File: internal/api/sms_provider/msg91.go (hunk 8 above)
Comment (MeetChaudhari) [reply internal/api/sms_provider/msg91.go:114]: ok

File: .github/workflows/test-m.yml (hunk 1 above)
Comment (kangmingtay) [reply .github/workflows/test-m.yml:1]: we shouldn't be merging this file in, can you please remove it?

File: example.docker.env
Comment (kangmingtay) [inline example.docker.env:1]: why is this file being deleted?

File: example.env
Comment (kangmingtay) [inline example.env:1]: ditto, we shouldn't be deleting this

**Review by kangmingtay — CHANGES_REQUESTED** [review]

File: internal/api/sms_provider/msg91.go (hunk 2 above)
Comment (hf) [reply internal/api/sms_provider/msg91.go:50]: Hey @MeetChaudhari we are not going to be able to merge this PR with this parameter. Please remove it!

File: internal/api/sms_provider/msg91.go (hunk 4 above)
Comment (kangmingtay) [reply internal/api/sms_provider/msg91.go:96]: can you please also provide the link to their API documentation so we can take a look at the params being passed in and what they mean?

File: internal/conf/configuration.go [hunk 11]
```diff
@@ -217,6 +218,12 @@ type TextlocalProviderConfiguration struct {
 	Sender string `json:"sender" split_words:"true"`
 }
 
+type Msg91ProviderConfiguration struct {
+	AuthKey       string `json:"auth_key" split_words:"true"`
+	SenderId      string `json:"sender_id" split_words:"true"`
+	DltTemplateId string `json:"dlt_template_id" split_words:"true"`
```
Comment (kangmingtay) [inline internal/conf/configuration.go:224]: i'm not familiar with what a DltTemplateId is - can you please link the relevant documentation for this? is it a mandatory field or can you send messages through msg91 without it?

File: example.env [hunk 12]
```diff
@@ -1,211 +0,0 @@
-# General Config
-# NOTE: The service_role key is required as an authorization header for /admin endpoints
-
-GOTRUE_JWT_SECRET="CHANGE-THIS! VERY IMPORTANT!"
```
Comment (J0) [inline example.env:4]: /nit When this file is re-introduced, do add the example env vars for both `example.docker.env` and `example.env` 



e.g. 
```
GOTRUE_SMS_MSG91_API_KEY=""
GOTRUE_SMS_MSG91_SECRET=""
GOTRUE_SMS_VONAGE_FROM=""
```

etc

File: internal/api/sms_provider/msg91.go (hunk 2 above)
Comment (kangmingtay) [reply internal/api/sms_provider/msg91.go:50]: Yeah @MeetChaudhari, gotrue is meant to work independently from Supabase. Some of our users self-host gotrue and don't use Supabase but might want to use msg91 as an sms platform. I don't think it's a good idea to add a tracking param here and i don't think our users would appreciate that as well.

**Comment by J0** [comment]
Thanks for the work on the PR thus far! Gentle reminder to file a PR to the [phone-login guide](https://github.com/supabase/supabase/tree/cff623e0107b71929bece8918b91bff7bbe5e381/apps/docs/pages/guides/auth/phone-login) with setup instructions once done so that we can try to test it. It'll also be really helpful in helping other developers figure out how to use the integration.

File: internal/api/sms_provider/msg91.go (hunk 4 above)
Comment (MeetChaudhari) [reply internal/api/sms_provider/msg91.go:96]: https://api.msg91.com/apidoc/textsms/send-sms-v1.php

File: internal/api/sms_provider/msg91.go (hunk 2 above)
Comment (MeetChaudhari) [reply internal/api/sms_provider/msg91.go:50]: I agree and i talked with msg91 team but this is the response form the msg91 team - 
"This is used to keep a track of requests coming in from Supabase. In case any error occurs, we  will be able to track the request quickly to provide a better experience.

This is majorly used to manage queues at our end and track any ongoing errors. 

It doesn't store or imply any customer details.
It just gives us report of total SMS failed(not user wise) with this Pluginsource so we can work on reducing number of failures and improve efficiency." 

what should i do?

File: internal/api/sms_provider/msg91.go (hunk 7 above)
Comment (MeetChaudhari) [reply internal/api/sms_provider/msg91.go]: I reimplemented it in different way, please check it out

File: internal/api/sms_provider/msg91.go (hunk 2 above)
Comment (kangmingtay) [reply internal/api/sms_provider/msg91.go:51]: Unfortunately, you'll have to remove this one. If there are any issues with Msg91's service, we'll just ask the customer to reach out the Msg91 directly.

Comment (MeetChaudhari) [reply internal/api/sms_provider/msg91.go:51]: ok i'll remove that.

File: internal/conf/configuration.go (hunk 10 above)
Comment (MeetChaudhari) [reply internal/conf/configuration.go]: Can you explain, like should i check for string type ?. Sorry i didnt understand

File: example.docker.env
Comment (MeetChaudhari) [reply example.docker.env:1]: i guess now its not getting deleted

**Review by hf — COMMENTED** [review]
This is generally good now @MeetChaudhari, @kangmingtay, @J0. I'll try to fix some of the technical comments and merge it.

**Comment by MeetChaudhari** [comment]
any update on this?

**Comment by hf** [comment]
@MeetChaudhari I've forked this #1018 and fixed it up. I'll close this PR as it's quite out of date. 

Do you think it's possible for you to contribute some docs for Msg91 similar to these here: https://supabase.com/docs/guides/auth/phone-login/twilio

**Comment by MeetChaudhari** [comment]
> @MeetChaudhari I've forked this #1018 and fixed it up. I'll close this PR as it's quite out of date. 
> 
> Do you think it's possible for you to contribute some docs for Msg91 similar to these here: https://supabase.com/docs/guides/auth/phone-login/twilio

Yes i can do it, actually I was waiting for pr to get merged
